src/foam/nanos/blob/FileService.java

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.blob;
+
+import foam.blob.Blob;
+import foam.blob.BlobService;
+import foam.blob.IdentifiedBlob;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.fs.File;
+import org.apache.commons.io.IOUtils;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.OutputStream;
+
+public class FileService
+    extends HttpBlobService
+{
+  protected DAO fileDAO_;
+  protected DAO userDAO_;
+
+  public FileService(X x, BlobService delegate) {
+    super(x, delegate);
+    fileDAO_ = (DAO) x.get("fileDAO");
+    // use the user dao instead of local user dao
+    // so that we get the authentication decoration
+    userDAO_ = (DAO) x.get("userDAO");
+  }
+
+  @Override
+  protected void download(X x) {
+    OutputStream os = null;
+    HttpServletRequest  req  = x.get(HttpServletRequest.class);
+    HttpServletResponse resp = x.get(HttpServletResponse.class);
+
+    try {
+      String path = req.getRequestURI();
+      String id = path.replaceFirst("/service/" + nspec_.getName() + "/", "");
+
+      // find file from file dao
+      File file = (File) fileDAO_.find_(x, id);
+      if ( file == null || !(file.getData() instanceof IdentifiedBlob) ) {
```

Comment:
Let's log a different error if file.getData() is not a blob.

That sounds more like an internal service error.

Reply:
I'm modifying this if statement to just check if file.getData == null. In this case the 404 status still makes sense.